### PR TITLE
Make joining decision pluggable, #144

### PR DESCRIPTION
**Pull request by patriknw**
* This centralized decision point also makes the actors easier to understand
* This probably fixes some issues with the old implementation also

Refs #144

### CODE REVIEW

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/JoinDecider.scala [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+              "Exceeded stable margins but missing seed node information from some contact points [{}] (out of [{}])",
+              contactPointsWithoutSeedNodesObservations.mkString(", "), info.contactPoints.mkString(", "))
+
+        KeepProbing.asCompletedFuture
+      }
+
+    }
+
+  // FIXME Do we really need settings.contactPoint.noSeedsStableMargin? Isn't the stable-margin enough?
+  // It has been probing all known contact points for that duration
```
[the remark this answers is not in the document]
Comment (ktoso) [reply cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/JoinDecider.scala]: Yes I think you're right; those two can be collapsed into just using stable-margin for the decision point

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala [hunk 2]
```diff
@@ -279,16 +243,27 @@ final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDiscovery, setti
...
+
+      // filter out old observations, in case the probing failures are not triggered
+      def isObsolte(obs: SeedNodesObservation): Boolean =
+        Duration.between(obs.observedAt, currentTime).toMillis > settings.contactPoint.probingFailureTimeout.toMillis
+
+      val seedObservations = seedNodesObservations.valuesIterator.filterNot(isObsolte).toSet
+      val info =
+        new SeedNodesInformation(currentTime, contacts.observedAt, contacts.observedContactPoints, seedObservations)
+
+      joinDecider.decide(info).recover { case _ => KeepProbing }.foreach(self ! _)
```
Comment (ktoso) [inline cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: recover is good, but also log a warning then?

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala [hunk 3]
```diff
@@ -135,130 +103,126 @@ final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDiscovery, setti
   import HeadlessServiceDnsBootstrap._
   import context.dispatcher
 
+  // FIXME move/rename these classes, doesn't have to be DNS
```
Comment (ktoso) [inline cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: yes good point Neither is it about Headless now as well, which is nice I think...? Headless is a specific k8s term, but all this is very generic now :)

**Comment by ktoso** [comment]
Resolves https://github.com/akka/akka-management/issues/133 as well

File: cluster-bootstrap/src/main/resources/reference.conf [hunk 4]
```diff
@@ -74,7 +74,7 @@ akka.management {
 
       # If some discovered seed node will keep failing to connect for specified period of time,
       # it will initiate rediscovery again instead of keep trying.
-      probing-failure-timeout = 5 seconds
+      probing-failure-timeout = 3 seconds
```
Comment (ktoso) [inline cluster-bootstrap/src/main/resources/reference.conf]: I think this needs some cleanup... The request coming back does not have a timeout it seems; yet this timeout is about "if we KEEP FAILING for this period, we'll resolve a new contact point and try again". I think we need to add a probe-timeout = 1.second perhaps and then should be good to have this at 3, so it's be 2/3 attempts after which we resolve again

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala (hunk 2 above)
Comment (ktoso) [reply cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: I guess no timeout is fine here hm; Technically we could end up invoking it again while it's still internally working on the previous decision future -- could lead to conflicting decisions? I guess we'd have acted upon the first thing already but maybe could show up as confusing logs if the decider decides to log things. 

Some more guard rails around this case would be nice... maybe if only logging that previous future decision still not completed etc

**Comment by patriknw** [comment]
Incorporated some of the feedback, in 56ec7ac
Will fix the other things also.

**Comment by patriknw** [comment]
@ktoso Incorporated remaining feedback and updated documentation. This is ready for final review.

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala [hunk 5]
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
       }
   }
 
-  private def decide(): Unit =
-    lastContactsObservation.foreach { contacts =>
-      val currentTime = timeNow()
+  private def decide(): Unit = {
+    if (decisionInProgress)
+      log.debug("Previous decision still in progress")
```
Comment (ktoso) [inline cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: Good though I was thinking warning; but debug is fine I guess

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala [hunk 6]
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
...
-        new SeedNodesInformation(currentTime, contacts.observedAt, contacts.observedContactPoints, seedObservations)
+        decisionInProgress = true
 
-      joinDecider.decide(info).recover { case _ => KeepProbing }.foreach(self ! _)
+        joinDecider
+          .decide(info)
+          .recover {
+            case e =>
+              log.error(e, "Join decision failed: {}", e)
+              KeepProbing
```
Comment (ktoso) [inline cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: good :)

File: cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala [hunk 7]
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
...
-      val info =
-        new SeedNodesInformation(currentTime, contacts.observedAt, contacts.observedContactPoints, seedObservations)
+        decisionInProgress = true
 
-      joinDecider.decide(info).recover { case _ => KeepProbing }.foreach(self ! _)
+        joinDecider
+          .decide(info)
+          .recover {
+            case e =>
+              log.error(e, "Join decision failed: {}", e)
```
Comment (ktoso) [inline cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala]: e.getMessage) likely? Will change

**Comment by patriknw** [comment]
@ktoso I fixed one thing that I had missed. `form-new-cluster` was lost in the refactoring, fixed in 3a6c9ee

I also moved the `LowestAddressJoinDecider` to a separate file as you suggested previously.

I'll do the other renaming of the classes in a separate PR, don't want to mix that with this diff.

**Comment by ktoso** [comment]
Very nice, thanks!

**Pull request merged**